asm/instruction_test.go

```diff
@@ -280,3 +288,46 @@ func TestInstructionIterator(t *testing.T) {
 		}
 	}
 }
+
+func TestMetadataCopyOnWrite(t *testing.T) {
+	insn := Instructions{
+		JEq.Imm(R1, 123, "my_func"),
+		Mov.Reg(R2, R1),
+		Mul.Reg(R1, R2),
+	}
+
+	insn2 := make([]Instruction, len(insn))
+	copy(insn2, insn)
+
+	// Setting the reference should have copied the metadata object and not effected the exiting pointer
+	insn2[0].SetReference("my_func2")
+
+	if insn[0].Reference() != "my_func" {
+		t.Fatal("metadata modification to copied instruction effected the old instruction")
+	}
+
+	if insn2[0].Reference() != "my_func2" {
+		t.Fatal("SetReference didn't update new instruction")
+	}
+
+	if insn[0].metadata == insn2[0].metadata {
+		t.Fatal("changed instruction should not be equal")
+	}
+
+	// Set the reference, then clear it. Causing us to have a nil in one instruction and an empty metadata in the
+	// other.
+	insn[1].SetReference("SomeValue")
+	insn[1].SetReference("")
```

Comment:
I think this sets the reference on the wrong insn.

Reply:
This is intentional, since this will result in an allocated metadata object with an empty string. Testing equality in such an edge case. But this is no longer important since we don't compare metadata anymore in `equal()`, so we can ditch it.